# SVM (Máquinas de Vetores de Suporte)

In [1]:
from sklearn.svm import SVC
import pandas as pd
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score


## Importando dados

In [2]:
Y_treino = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/Y_treino.csv', sep=',', encoding='utf-8')
Y_teste = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/Y_teste.csv', sep=',', encoding='utf-8')

X_esc_treino = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/X_esc_treino.csv', sep=',', encoding='utf-8')
X_esc_teste = pd.read_csv('/home/lucasqueiros/DataScience/Projeto-Classificacao/data/processed/X_esc_teste.csv', sep=',', encoding='utf-8')

In [3]:
Y_teste = Y_teste.iloc[:, 1].values
Y_treino = Y_treino.iloc[:, 1].values

X_esc_teste = X_esc_teste.iloc[:, 1:].values
X_esc_treino = X_esc_treino.iloc[:, 1:].values

## Treinamento do Algoritmo com X

In [5]:
svm1 = SVC(kernel='linear', C=7, random_state=42)
svm1.fit(X_esc_treino, Y_treino)

SVC(C=7, kernel='linear', random_state=42)

In [6]:
previsoes_svm1 = svm1.predict(X_esc_teste)
previsoes_svm1

array([0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0])

In [7]:
print("Acurácia: %.2f%%" % (accuracy_score(Y_teste, previsoes_svm1) * 100.0))

Acurácia: 97.66%


- svm1 = SVC(kernel='rbf', C=5, random_state=1) - 95.32%
- svm1 = SVC(kernel='linear', C=7, random_state=1) - 97.66%

In [8]:
confusion_matrix(Y_teste, previsoes_svm1)

array([[106,   2],
       [  2,  61]])

### Análise com dados de treino

In [9]:
previsoes_treino1 = svm1.predict(X_esc_treino)
previsoes_treino1

array([0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1,
       0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,

In [10]:
print("Acurácia: %.2f%%" % (accuracy_score(Y_treino, previsoes_treino1) * 100.0))

Acurácia: 96.73%


In [11]:
confusion_matrix(Y_treino, previsoes_treino1)

array([[246,   3],
       [ 10, 139]])

## Validação Cruzada

In [12]:
# Separando os dados em 30 partes
kfold = KFold(n_splits=30, shuffle=True, random_state=42)

In [13]:
# Criando o modelo
modelo = SVC(kernel='linear', C=7, random_state=42)
resultado = cross_val_score(modelo, X_esc_treino, Y_treino, cv=kfold)
resultado

array([1.        , 0.85714286, 0.92857143, 1.        , 1.        ,
       0.92857143, 0.78571429, 1.        , 0.84615385, 0.92307692,
       1.        , 0.92307692, 0.92307692, 1.        , 1.        ,
       1.        , 0.84615385, 0.84615385, 1.        , 1.        ,
       0.84615385, 0.76923077, 1.        , 1.        , 0.92307692,
       1.        , 1.        , 1.        , 1.        , 1.        ])

In [14]:
# Usamos a média e o desvio padrão
print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 94.49%


##### SVM - 97.66% (treino e teste) e 94.49% (validação cruzada) - previsores - SVC(kernel='linear', C=7, random_state=42)